File: games/hangman.go
package games

import (
	"fmt"
	"strings"
	"unicode"
)

func createBlank(s string, guesses map[string]bool) string {
	var blanks string

	for _, i := range s {
		_, ok := guesses[string(unicode.ToLower(i))]
		if ok {
			blanks += string(i) + " "
		} else {
			if unicode.IsLetter(i) {
				blanks += "_ "
			} else {
				blanks += string(i) + " "
			}
		}
	}
	return blanks
}

func hasWon(s string, guesses map[string]bool) bool {
	for _, i := range s {
		if unicode.IsLetter(i) {
			_, ok := guesses[string(unicode.ToLower(i))]
			if !ok {
				return false
			}
		}
	}
	return true
}

func formatGuesses(guesses map[string]bool) string {
	str := "Guesses [ "
	alphabet := "abcdefghijklmnopqrstuvwxyz"
	for _, i := range alphabet {
		_, ok := guesses[string(i)]
		if ok {
			str += string(i) + " "
		}
	}
	return str + "]"
}

func formatHangman(lives int) string {
	var head string
	var torso string
	var legs string

	if lives <= 5 {
		head = " O "
	}

	if lives <= 4 {
		switch lives {
		case 4:
			torso = " | "
		case 3:
			torso = "/| "
		default:
			torso = "/|\\"
		}
	}

	if lives <= 1 {
		switch lives {
		case 1:
			legs = "/  "
		default:
			legs = "/ \\"
		}
	}

	return fmt.Sprintf("    |==|\n    | %s\n    | %s\n    | %s\n  __|__\n", head, torso, legs)

	// |---|
	// |   O
	// |  /|\
	// |  / \
	// |_____
}

func Hangman() Result {
	lives := 6
	guesses := map[string]bool{}
	_, _ = lives, guesses
	word := RandomWord()

	for lives > 0 && !hasWon(word, guesses) {
		blanks := createBlank(word, guesses)

		fmt.Println(formatGuesses(guesses))
		fmt.Println()
		fmt.Printf(formatHangman(lives))
		fmt.Println(blanks)

		guess := GetUserString()

		fmt.Println()

		if len(guess) != 1 {
			fmt.Println("Guess a letter!")
			fmt.Println()
			continue
		} else if !unicode.IsLetter(rune(guess[0])) {
			fmt.Println("Guess a letter!")
			fmt.Println()
			continue
		}

		lowerGuess := string(unicode.ToLower(rune(guess[0])))
		_, ok := guesses[lowerGuess]

		if ok {
			fmt.Printf("You already guessed %s", lowerGuess)
			fmt.Println()
			fmt.Println()
			continue
		}

		guesses[lowerGuess] = true

		if strings.Contains(word, lowerGuess) {
			fmt.Printf("%s is in the word.", lowerGuess)
		} else {
			fmt.Printf("%s is not in the word.", lowerGuess)
			lives--
		}
		fmt.Println()
		fmt.Println()
	}

	fmt.Println(formatHangman(lives))
	fmt.Printf("The word was: %s\n", word)
	if hasWon(word, guesses) {
		fmt.Println("You Win!")
		return Result{
			Points:    8,
			PlayerWin: true,
		}
	} else if lives <= 0 {
		fmt.Println("You Lose...")
		return Result{
			Points:    1,
			PlayerWin: false,
		}
	}
	return Result{}
}

File: src/games/hangman.rs
use crate::{games::*, *};

pub struct Hangman {
    word: String,
    max_guesses: usize,
    total_guesses: usize,
    letters_guessed: Vec<char>,
}

impl Hangman {
    pub fn new(word: String) -> Hangman {
        Hangman {
            word,
            max_guesses: 8,
            total_guesses: 0,
            letters_guessed: Vec::new(),
        }
    }

    //   ‖‾‾‾|
    //   ‖   O
    //   ‖  /|\
    //   ‖   |
    //   ‖  / \
    // --╨--
    fn format_hangman(&self) -> String {
        let head = match self.total_guesses {
            n if n == self.max_guesses => "   X",
            1.. => "   O",
            _ => "",
        };

        let torso = match self.total_guesses {
            2 => "   |",
            3 => "  /|",
            4.. => "  /|\\",
            _ => "",
        };

        let abdomen = match self.total_guesses {
            5.. => "   |",
            _ => "",
        };

        let legs = match self.total_guesses {
            6 => "  /",
            7.. => "  / \\",
            _ => "",
        };

        let top = String::from("  ‖‾‾‾|");
        let head_section = String::from("  ‖") + head;
        let torso_section = String::from("  ‖") + torso;
        let abdomen_section = String::from("  ‖") + abdomen;
        let leg_section = String::from("  ‖") + legs;
        let bottom = String::from("--╨--");

        format!(
            "{}\n{}\n{}\n{}\n{}\n{}",
            top, head_section, torso_section, abdomen_section, leg_section, bottom
        )
    }

    fn blank_unguessed(&self) -> String {
        let mut blanks = String::new();
        for letter in self.word.chars() {
            if is_letter(letter) && !self.letters_guessed.contains(&letter.to_ascii_lowercase()) {
                blanks += "_ ";
            } else {
                blanks += &format!("{} ", letter);
            }
        }
        blanks
    }

    fn unique_letters(&self) -> Vec<char> {
        let mut unique = Vec::new();
        for letter in self.word.chars() {
            if !unique.contains(&letter) && is_letter(letter) {
                unique.push(letter.to_ascii_lowercase());
            }
        }
        unique
    }

    fn has_won(&self, letters: Vec<char>) -> bool {
        for letter in letters {
            if !self.letters_guessed.contains(&letter) {
                return false;
            }
        }
        true
    }
}

impl Game for Hangman {
    fn gameloop(&mut self) -> GameResult {
        while self.total_guesses < self.max_guesses {
            println!();
            println!("{}", self.format_hangman());
            println!("{}", self.blank_unguessed());
            println!(
                "Guessed: [{}]",
                self.letters_guessed
                    .iter()
                    .map(|c| c.to_string())
                    .collect::<Vec<String>>()
                    .join(" ")
            );
            println!(
                "Guesses Remaining: {}",
                self.max_guesses - self.total_guesses
            );
            let letters_in_word = self.unique_letters();

            let mut letter: char;

            // prompt user until a valid guess is entered
            loop {
                // get input on the same line
                println!();
                print!("Guess a letter: ");
                io::stdout().flush().unwrap();

                let guess = user_string();
                letter = match first_char(&guess) {
                    Some(c) => c.to_ascii_lowercase(),
                    None => {
                        // println!("Please guess a letter.");
                        println!("{}", "Please guess a letter.".red());
                        continue;
                    }
                };

                // check if guess is a valid letter
                if !is_letter(letter) {
                    // println!("'{}' is not a letter!", guess);
                    println!("{}", format!("'{}' is not a letter!", guess).red());
                    continue;
                }

                // check if letter has already been guessed
                if self.letters_guessed.contains(&letter) {
                    println!("{}", format!("You already guessed '{}'!", letter).red());
                    continue;
                }

                // add letter to guessed
                self.letters_guessed.push(letter);

                break;
            }

            if letters_in_word.contains(&letter) {
                println!("{}", format!("'{}' is in the word!", letter).green());
            } else {
                self.total_guesses += 1;
                println!("{}", format!("'{}' is not in the word!", letter).red());
            }

            if self.has_won(letters_in_word) {
                println!();
                println!("{}", self.format_hangman());
                println!("{}", self.blank_unguessed());
                println!("{}", "You Win!".green());
                println!("The word was: '{}'", self.word);
                return GameResult {
                    winner: Winner::Player,
                    award: self.total_guesses as i32,
                };
            }
        }
        println!();
        println!("{}", self.format_hangman());
        println!("You Lose!");
        println!("The word was: {}", self.word);
        GameResult {
            winner: Winner::Amy,
            award: (self.max_guesses / 2) as i32,
        }
    }
}
